17642 adds edit state to frames to fix new tab issue

anthonyburchell opened the pull request:
## Description
https://core.trac.wordpress.org/ticket/48028
Excerpt from the ticket above: 

Fixes: #17642.

> The "Media Library" panel of the "Featured Image" modal dialog contains a link labelled with "Edit image". This link takes users to a new window/tab, but that link does not inform users that it opens a new window or tab.
A link does not naturally convey itself to assistive technology that it opens in a new tab without link text that describes it.
It is important to convey that a link opens in a new tab because of the change of context may not be clear to everyone.

<img width="639" alt="Screen Shot 2019-09-29 at 4 35 55 PM" src="https://user-images.githubusercontent.com/11888980/65839906-71680800-e2d7-11e9-8eb2-90bfe8462007.png">


## ## Types of changes/ How has this been tested?

This PR adds a new editState function to initialize the available edit frames. This function also respects the `allowedTypes` prop and has been tested with the video, gallery, and featured image media flows. Additional testing there would be appreciated. 

~~PENDING BUG: edit image state for gallery seems to be messing up because of the multiple nature of this screen. Leaving this draft until I have a fix.~~

Commit:
adds edit state to frames to fix new tab issue

Commit:
moves default state build into else of gallery condition and adds gallery toolbar logic to reach parity with Core.

Commit:
fixes selection object reference to proper state

Commit:
improves and adds documentation

joemcgill commented on packages/media-utils/src/components/media-upload/index.js:
```diff
@@ -86,6 +165,43 @@ const getGalleryDetailsMediaFrame = () => {
+ * @param {Object} types passes the allowedTypes object to the media library.
+ * @return {wp.media.view.MediaFrame.Select} The default media workflow.
+ */
+const addEditState = ( types ) => {
```
Curious if we could move this into core directly, rather than extending the Select frame only for Gutenberg to include the edit state?

anthonyburchell replied:
Thanks for looking at this! I explored doing it globally in the select frame but found it difficult to set the context of `allowedTypes` and selections from this component. The way that we are using frames in Gutenberg is a bit different than we do with core because we base everything off of one default frame. If we set it globally in the select frame, we would have conflict with the Video block because we use the default frame with mime types passed from the props. In this function, we need to pass the allowed mimetypes to the `createStates` function as `types` and as such, each frame (featured, gallery, default) has a specific state to get the selection from. (library, gallery, featured-image) I think where I found the hangup is passing `allowedTypes` from Gutenberg to the Core code. I could be missing something so please let me know if you have any ideas how :) I also want to be cautious about changing Core code and breaking the Classic Editor plugin's flow

anthonyburchell replied:
Have it working in the Core `Select` ! I need to explore one last bug where the select button is blank when clicking "back" from the edit screen. This addEditState function can now be removed from this PR. The featured image and gallery states will still need this PR to fix it in Gutenberg. Thanks for calling this out @joemcgill !

gziolo replied:
> Have it working in the Core `Select` !

This is great news! Did you manage to resolve it in Core fully?

anthonyburchell replied:
@gziolo Yeah! I think we are good now. I've updated this PR to remove the default edit state handling. Now the edit state is added to select using the latest patch in https://core.trac.wordpress.org/ticket/48028

With the core patch above **and** this PR applied to a local Gutenberg, the issue is fixed in gallery, featured image, image block, and inline image block. (may npm have mercy on your computer while testing)

I also removed a declaration in the `frameConfig` to not define a button text for the default state. It is redundant with the core default select button text. Also it caused conflict when hitting the back button once in the `Edit Image` frame.

anthonyburchell replied:
and to be clear. The Core side is resolved fully once the patch is committed, however, this PR is needed in order for folks using the new editor to reap the benefits. :) 

New commit incoming to resolve Travis errors.

Commit:
removes buildAndSetEditFrame as it should be added in core and not here. Also removes the declaration of the button text as the default already is 'select'

Commit:
truly removes editState fu function

anthonyburchell commented:
The Core side changes have been merged in https://core.trac.wordpress.org/changeset/46461

No need to apply the Core patch now. Just test this PR and all should be working again.

gziolo commented:
I tested it with the latest WordPress core changes in `trunk`. There is some edge case where you see only a blank page in the modal after clicking on the `Edit Image` link. I'm not 100% sure how to reproduce it, but it feels like it happens after you close the modal on the `Edit Image` page and you try to edit the same image from another part of the UI:

![edit-image](https://user-images.githubusercontent.com/699132/66749026-072f9580-ee89-11e9-879b-fbbe13667ab0.gif)

I also can observe that there is some unexpected behavior with the focus when navigating back from editing the image when you want to select the featured image:

![Screen Shot 2019-10-14 at 13 23 31](https://user-images.githubusercontent.com/699132/66749128-4827aa00-ee89-11e9-965a-5c2479728d8c.png)

Somehow, the visual focus stays on the `Back` button which is no longer there.

joemcgill commented on packages/media-utils/src/components/media-upload/index.js:
```diff
@@ -49,14 +75,66 @@ const getGalleryDetailsMediaFrame = () => {
+		 * @return {void}
+		 */
+		editState( ) {
+			const selection = this.state( 'gallery' ).get( 'selection' ),
```
Same style suggestions from above.

joemcgill commented on packages/media-utils/src/components/media-upload/index.js:
```diff
@@ -26,21 +31,42 @@ const getFeaturedImageMediaFrame = () => {
+		 * @return {void}
+		 */
+		editState( ) {
+			const selection = this.state( 'featured-image' ).get( 'selection' ),
```
Nitpick, but I'd prefer separate `const` declarations for each variable rather than the shorthand because it makes it easier to read at a glance. I'd also move `this.content.set(view);` to a new line to separate functional code from the variable declarations and add an inline comment.

gziolo replied:
Yes, variable declarations are usually in their own statements in Gutenberg. I'm familiar with both approaches because of previous experience, but I agree that we could align.

joemcgill commented on packages/media-utils/src/components/media-upload/index.js:
```diff
@@ -26,21 +31,42 @@ const getFeaturedImageMediaFrame = () => {
+				view = new wp.media.view.EditImage( { model: selection.single(), controller: this } ).render();
+			this.content.set( view );
+
+			// after bringing in the frame, load the actual editor via an ajax call
```
Capitalize "After" and end the sentence with a period.

joemcgill reviewed (approved):
A few small stylistic comments, but the code overall looks good to me.

gziolo commented:
@anthonyburchell - we should figure out why you see this white screen inside the modal from time to time before we proceed.

There is also this question, whether we should remove `target="_blank"` from `Edit Image` link as it seems to be no longer necessary.

afercia commented:
The related Trac ticket is milestoned for WordPress 5.4. See https://core.trac.wordpress.org/ticket/48028

Any change to get this rebased, reviewed, and if everything goes well, merged? /Cc @youknowriad @anthonyburchell  @joemcgill

anthonyburchell committed:
rebase

anthonyburchell committed:
fixes linting issues

anthonyburchell commented:
Thanks for the ping @afercia! I rebased and fixed some linting issues. Looks like that is passing now. 

The feedback provided in [this comment](https://github.com/WordPress/gutenberg/pull/17641#issuecomment-541630454) is still valid. It is possible to get a blank white screen after clicking edit in one frame (gallery) and opening edit in another (maybe featured). I still have some investigation to get that solved. It's a pretty edge case instance so I leave it to y'all to thumbs up moving forward and making a separate issue to solve that.

anthonyburchell commented:
Also, it may be worth mentioning that I don't think this _needs_ to be included in Beta 1. This is a bug fix to a broken part of the edit image functionality so maybe we can land this for Beta 2?

anthonyburchell committed:
Adjusts PR to factor in review suggestions

anthonyburchell commented:
An update here, I found that the edit image bug has existed since before this PR. I think that calls for a trac side change to fix the close action. To reproduce, click edit image on a selected image in the gallery library modal, close window. open another frame (featured image maybe), attempt to edit same image, see white screen. It's possible to do this in older versions. 

With that in mind, I dont see any blockers to merging this PR before 5.4 Beta 2. Please let me know if theres anything I can do to bring it to 100%. :)

gziolo reviewed (approved):
I tested it again and I couldn't reproduce issues listed a few months back. Nice work. Let's get it in 🎉

gziolo commented:
> With that in mind, I dont see any blockers to merging this PR before 5.4 Beta 2. Please let me know if theres anything I can do to bring it to 100%. :)

@jorgefilipecosta - I added this PR to the list of PRs to include in WP 5.4 release.

mcsf commented:
@anthonyburchell @gziolo @jorgefilipecosta:
There are two bugs introduced in the Gallery block by this PR:

## Gallery: can't insert select items from media library

![gallery-media-insert](https://user-images.githubusercontent.com/150562/74732062-4f006800-5241-11ea-80b1-5151df48603c.gif)

Also notice that, when an image is selected in the media modal, the action button at the bottom-right corner reads **"Insert gallery"** instead of "Create a new gallery".

## Gallery: can't edit collection of images

1. Open a post that has a Gallery block with images.
2. Select the block.
3. Click on "Media Library" underneath the images.
**Expected:** The media modal opens up.
**Observed:** Nothing happens in the UI, and a TypeError is raised:

```
TypeError: t is null                      media-views.min.js:2:7343
```

anthonyburchell commented:
@mcsf you rock. Thank you for finding this. Here is the fix for the issue you identified: https://github.com/WordPress/gutenberg/pull/20287

Would greatly appreciate some testing on this to verify gallery works as intended. 🙏🏼
